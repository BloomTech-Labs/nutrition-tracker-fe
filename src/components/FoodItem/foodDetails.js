import moment from "moment";
import React, { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Textfit } from "react-textfit";
import { useToasts } from "react-toast-notifications";
import {
  DropdownItem as BS_DropdownItem,
  ButtonDropdown,
  DropdownMenu
} from "reactstrap";
import styled from "styled-components";
import { faCheck } from "@fortawesome/free-solid-svg-icons";
import {
  addFoodItem,
  getOneFoodItem
} from "../../store/actions/foodItemAction";
import Loading from "../Global/Loading";
import MacroBudgets from "../Global/MacroBudgets";
import Flywheel from "../Global/flywheel-menu/Flywheel";
import { Col, DropdownToggle, H2, H3, Input, Row } from "../Global/styled";
import NutritionInfo from "./components/NutritionInfo";

const FoodDetails = props => {
  const dispatch = useDispatch();
  const { addToast } = useToasts();

  const { consumed, currentDate, currentTimeZone } = useSelector(
    state => state.dailyLog
  );
  const { item, got, error } = useSelector(state => state.foodItemsReducer);
  const firebaseID = useSelector(state => state.firebase.auth.uid);

  const [quantity, setQuantity] = useState(1);
  const [dropdownOpen, setDropdownOpen] = useState(false);
  const [dropDownSelectionIndex, setDropDownSelectionIndex] = useState(0);

  // to populate date input
  const todayDate = moment.tz(currentTimeZone).format("YYYY-MM-DD");

  // to populate time input
  const todayTime = moment.tz(currentTimeZone).format("HH:mm");

  const [time, setTime] = useState(todayTime);
  const [date, setDate] = useState(todayDate);
  const [dateTimeUTC, setDateTimeUTC] = useState(
    moment
      .tz(`${date} ${time}`, currentTimeZone)
      .utc()
      .format()
  );

  useEffect(() => {
    setDateTimeUTC(
      moment
        .tz(`${date} ${time}`, currentTimeZone)
        .utc()
        .format()
    );
  }, [date, time]);

  useEffect(() => {
    dispatch(getOneFoodItem(props.match.params.fatsecret_food_id));
  }, [props.match.params.fatsecret_food_id]);

  useEffect(() => {
    if (got)
      addToast("Food Item Added!", {
        appearance: "success",
        autoDismiss: true
      });
    else if (error)
      addToast("Error. Try again later.", {
        appearance: "error",
        autoDismiss: true
      });
  }, [error, got]);

  const handleToggle = e => {
    e.preventDefault();
    setDropdownOpen(!dropdownOpen);
  };

  const handleSelect = key => {
    setDropDownSelectionIndex(key);
  };

  const addedMacros = () => {
    const selectionIndex = dropDownSelectionIndex;
    /* ****************************************************** */
    const addedfatGrams = Number(item[selectionIndex].fat_g) * quantity;
    const addedCarbGrams = Number(item[selectionIndex].carbs_g) * quantity;
    const addedProteinGrams = Number(item[selectionIndex].protein_g) * quantity;

    // rounds to nearest hundreth
    return {
      fat: Math.round(100 * addedfatGrams) / 100,
      carbs: Math.round(100 * addedCarbGrams) / 100,
      protein: Math.round(100 * addedProteinGrams) / 100
    };
  };

  const addNewFoodLog = async () => {
    const selectionIndex = dropDownSelectionIndex;
    /* ****************************************************** */
    const food_id = item[selectionIndex].id;
    const serving_id = item[selectionIndex].serving_id;
    const fatsecret_food_id = props.match.params.fatsecret_food_id;
    const time_consumed_at = dateTimeUTC;
    const daily_nutrition_totals_date = moment
      .tz(dateTimeUTC, currentTimeZone)
      .format("YYYY-MM-DD");
    const time_zone_name = currentTimeZone;
    const time_zone_abbr = getCurrentTimeZoneAbbr();

    dispatch(
      addFoodItem(
        {
          food_id,
          quantity,
          serving_id,
          fatsecret_food_id,
          time_consumed_at,
          daily_nutrition_totals_date,
          time_zone_name,
          time_zone_abbr
        },
        firebaseID
      )
    );
  };

  const getCurrentTimeZoneAbbr = () => {
    const time_zone_abbr = moment.tz(currentDate, currentTimeZone).format("z"); // output ex. PST

    return time_zone_abbr;
  };

  if (!item[0]) return <Loading />;

  /* ****************************************************** */
  const foodSelection = item[dropDownSelectionIndex];

  return (
    <div>
      <Row>
        <Col
          align="center"
          height="50px"
          style={{ borderBottom: "1px solid lightgrey" }}
        >
          <FoodName id="food-name">
            <Textfit mode="single" forceSingleModeWidth={false}>
              {foodSelection && foodSelection.food_name}
            </Textfit>
          </FoodName>
        </Col>
      </Row>
      <Row style={{ paddingTop: "20px" }}>
        <Col align="center" height="50px">
          <CurrentCalories>
            Current Cal
            <br />
            <span>{consumed.caloriesConsumed} cal</span>
          </CurrentCalories>
        </Col>
        <Col align="center" height="50px">
          <AddedCalories>
            {""} <br />
            <span>
              {" "}
              +
              {Math.trunc(
                foodSelection && foodSelection.calories_kcal * quantity
              )}{" "}
              cal
            </span>
          </AddedCalories>
        </Col>
        <Col align="center" height="50px">
          <NewCalories>
            New Cal
            <br />
            <span>
              {consumed.caloriesConsumed +
                foodSelection.calories_kcal * quantity}{" "}
              cal
            </span>
          </NewCalories>
        </Col>
      </Row>
      <MacroBudgets macrosAdded={addedMacros()} date={date} />
      <Row
        style={{
          marginTop: "50px",
          paddingTop: "15px",
          borderTop: "1px solid lightgrey"
        }}
      >
        <Col direction="column" align="flex-start">
          <InputLabel>Qty</InputLabel>
          <Input
            type="number"
            name="quantity"
            value={quantity}
            min={1}
            onChange={e => {
              setQuantity(e.target.value);
            }}
          />
        </Col>
        <Col direction="column" align="flex-end">
          <InputLabel>Serving Type</InputLabel>
          <ButtonDropdown
            isOpen={dropdownOpen}
            toggle={handleToggle}
            style={{ width: "100%" }}
          >
            <DropdownToggle
              caret
              style={{
                textAlign: "right",
                backgroundColor: "white",
                color: "black",
                borderColor: "#CED4DA"
              }}
            >
              {item[0] && foodSelection.serving_desc}
            </DropdownToggle>
            <DropdownMenu>
              {item.map((serving, key) => (
                <DropdownItem key={key} onClick={() => handleSelect(key)}>
                  {serving.serving_desc}
                </DropdownItem>
              ))}
            </DropdownMenu>
          </ButtonDropdown>
        </Col>
      </Row>
      <Row style={{ marginBottom: "35px" }}>
        <Col direction="column" align="flex-start">
          <InputLabel>Date</InputLabel>
          <Input
            type="date"
            name="date"
            value={date}
            style={{ textAlign: "left" }}
            onChange={e => setDate(e.target.value)}
            min={1}
          />
        </Col>
        <Col direction="column" align="flex-end">
          <InputLabel>Time</InputLabel>
          <Input
            type="time"
            name="time"
            value={time}
            style={{ textAlign: "left" }}
            onChange={e => setTime(e.target.value)}
            min={1}
          />
        </Col>
      </Row>

      <NutritionInfo foodSelection={foodSelection} quantity={quantity} />
      <Row>
        <Col>
          <Flywheel
            staticInitialButton
            onMainButtonClick={addNewFoodLog}
            maintButtonIcon={faCheck}
            childButtonIcons={[]}
          />
        </Col>
      </Row>
    </div>
  );
};

const DataCol = styled(Col)`
  margin-top: 2rem;
  margin-bottom: 8rem;
  h5 {
    margin-top: 0.4rem;
    font-size: 1.6rem;
  }
  h6 {
    font-size: 1.4rem;
  }
  .gray-box {
    width: 100%;
    height: 1rem;
    background: gray;
    border-bottom: 1px solid black;
  }
`;

const MainData = styled(Col)`
  border-bottom: 1px solid black;
  h5 {
    font-weight: bold;
  }
`;

const SubData = styled(Col)`
  border-bottom: 1px solid black;
  h5 {
    margin-left: 1rem;
  }
`;

const FoodName = styled(H2)`
  text-align: center;
  width: 100%;
  /* font-size: 2rem; */
  span {
    font-weight: 300;
  }
`;

const CurrentCalories = styled(H3)`
  width: 100%;
  text-align: center;

  span {
    font-weight: 300;
  }
`;

const AddedCalories = styled(H3)`
  width: 100%;
  text-align: center;

  span {
    font-weight: 300;
  }
`;

const NewCalories = styled(H3)`
  width: 100%;
  text-align: center;

  span {
    font-weight: 300;
  }
`;

const InputLabel = styled.span`
  font-size: 1.6rem;
`;

const DropdownItem = styled(BS_DropdownItem)`
  font-size: 1.6rem;
`;

export default FoodDetails;
