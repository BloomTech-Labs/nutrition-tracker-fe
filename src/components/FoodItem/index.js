import React from "react";
import { connect } from "react-redux";
import { Route } from "react-router-dom";
import { searchFoodItems } from "../../store/actions/foodItemAction";
import { Container } from "../Global/styled";
import FoodDetails from "./foodDetails";
import SearchPage from "./searchPage";

class FoodItem extends React.Component {
  constructor() {
    super();
    this.state = {
      searchTerm: ""
    };
  }
  render() {
    const { path } = this.props.match;
    
    return (
      <Container height={this.props.height} fluid>
        <Route
          path={`${path}/search`}
          exact
          render={props => <SearchPage {...props} path={path} />}
        />

        <Route
          path={`${path}/view/:fatsecret_food_id`}
          render={props => (
            <FoodDetails isLoading={this.props.isFetching} {...props} />
          )}
        />
      </Container>
    );
  }
}

const mapStateToProps = state => {
  return {
    items: state.foodItemsReducer.items,
    getting: state.foodItemsReducer.getting
  };
};

export default connect(mapStateToProps, { searchFoodItems })(FoodItem);
