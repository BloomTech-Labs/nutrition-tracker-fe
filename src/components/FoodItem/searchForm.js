import React from "react";
import { connect } from "react-redux";
import {
  Form,
  InputGroup,
  InputGroupAddon,
  InputGroupText,
  Row
} from "reactstrap";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { searchFoodItems } from "../../store/actions/foodItemAction";
import { Input } from "../Global/styled";

class SearchForm extends React.Component {
  constructor() {
    super();
    this.state = {
      searchTerm: ""
    };
  }

  handleSearch = e => {
    let input_value = e.target.value;
    this.setState(prevState => {
      return {
        ...prevState,
        searchTerm: input_value
      };
    });
    this.props.searchFoodItems(input_value);
  };

  render() {
    return (
      <Form style={{marginTop: "10px"}}>
        <Row form>
          <InputGroup>
            <InputGroupAddon addonType="prepend">
              <InputGroupText>
                <FontAwesomeIcon icon={["fas", "search"]} size="2x" />
              </InputGroupText>
            </InputGroupAddon>
            <Input
              value={this.state.searchTerm}
              onChange={this.handleSearch}
              type="text"
              name="search_term"
              id="search_term"
              placeholder="Search food item"
              bsSize="lg"
            />
          </InputGroup>
        </Row>
      </Form>
    );
  }
}

export default connect(null, { searchFoodItems })(SearchForm);

export { SearchForm };
