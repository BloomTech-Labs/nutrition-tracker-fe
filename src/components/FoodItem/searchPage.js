import React from "react";
import { connect } from "react-redux";
import { Col, Row } from "reactstrap";
import { getOneFoodItem } from "../../store/actions/foodItemAction";
import SearchForm from "./searchForm";
import SearchResults from "./searchResults";

class SearchPage extends React.Component {

  handleGetFoodItem = fatsecret_food_id => {
    this.props.getOneFoodItem(fatsecret_food_id).then(response => {
      if (response) {
        this.props.history.push(`${this.props.path}/view/${fatsecret_food_id}`);
      }
    });
  };

  render() {
    return (
      <>
        <Row>
          <Col>
            <SearchForm />
          </Col>
        </Row>
        <Row>
          <Col>
            {/* <TBody responsive hover size="lg"> */}
            <SearchResults handleGetFoodItem={this.handleGetFoodItem} />
            {/* </TBody> */}
          </Col>
        </Row>
      </>
    );
  }
}

export { SearchPage };
export default connect(null, { getOneFoodItem })(SearchPage);
