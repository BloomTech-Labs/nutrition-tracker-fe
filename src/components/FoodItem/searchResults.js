import React from "react";
import { connect } from "react-redux";
import { TBody } from "../Global/styled/";

class SearchResults extends React.Component {
  render() {
    return this.props.items ? (
      <TBody responsive hover size="lg">
        <tbody>
        {this.props.items.map((foodItem, key) => {
          return (
            <tr
              id="search_results"
              key={key}
              onClick={() => this.props.handleGetFoodItem(foodItem.food_id)}
            >
              <th scope="row" style={{textAlign: "left"}}> {foodItem.food_name} </th>
              <td style={{whiteSpace: "nowrap"}}>{foodItem.standard_amount}</td>
              <td style={{whiteSpace: "nowrap"}}>{foodItem.calories} cal</td>
            </tr>
          );
        })}
        </tbody>
      </TBody>
    ) : (
      <TBody /> // TODO: If this breaks, we may need <tbody></tbody> here
    );
  }
}

const mapStateToProps = state => {
  return {
    items: state.foodItemsReducer.items
  };
};

export { SearchResults };
export default connect(mapStateToProps, {})(SearchResults);
