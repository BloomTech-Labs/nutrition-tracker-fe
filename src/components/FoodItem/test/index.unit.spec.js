import { mount } from "enzyme";
import React from "react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router";
import { useToasts } from "react-toast-notifications";
import { ToastProvider } from "react-toast-notifications";
import sinon from "sinon";
import FoodItem from "../index.js";

let store;
let wrapper;

const mockItems = [
  {
    food_name: "Pupusa",
    standard_amount: 1,
    calories: 2.99
  },
  {
    food_name: "Tacos",
    standard_amount: 4,
    calories: 14.2
  },
  {
    food_name: "Cheese Hamburguer",
    standard_amountt: 1,
    calories: 2093.1
  }
];

const mockFoodItem = [
  {
        "id": 1,
        "fatsecret_food_id": 6847,
        "serving_id": 25124,
        "retrieved_at": "2019-12-15T06:14:38.021Z",
        "food_name": "Vegetable Soup (Dry Mix)",
        "serving_url": "https://www.fatsecret.com/calories-nutrition/generic/vegetable-soup-dry-mix-not-reconstituted?portionid=25124&portionamount=1.000",
        "serving_desc": "1 tbsp",
        "metric_serving_amt": "1.20",
        "metric_serving_unit": "g",
        "calories_kcal": "4.00",
        "fat_g": "0.06",
        "carbs_g": "0.72",
        "protein_g": "0.14",
        "saturated_fat_g": "0.03",
        "monounsaturated_fat_g": "0.02",
        "polyunsaturated_fat_g": "0.01",
        "trans_fat_g": null,
        "fiber_g": "0.00",
        "sugar_g": "0.05",
        "cholesterol_mg": "0.00",
        "sodium_mg": "81.00",
        "potassium_mg": "7.00",
        "vitamin_a_daily_pct": "0.00",
        "vitamin_c_daily_pct": "1.00",
        "calcium_daily_pct": "0.00",
        "iron_daily_pct": "0.00"
    },
    {
        "id": 2,
        "fatsecret_food_id": 6847,
        "serving_id": 25324,
        "retrieved_at": "2019-12-15T06:14:38.022Z",
        "food_name": "Vegetable Soup (Dry Mix)",
        "serving_url": "https://www.fatsecret.com/calories-nutrition/generic/vegetable-soup-dry-mix-not-reconstituted?portionid=25324&portionamount=1.000",
        "serving_desc": "1 packet",
        "metric_serving_amt": "38.50",
        "metric_serving_unit": "g",
        "calories_kcal": "125.00",
        "fat_g": "1.96",
        "carbs_g": "23.06",
        "protein_g": "4.52",
        "saturated_fat_g": "0.89",
        "monounsaturated_fat_g": "0.74",
        "polyunsaturated_fat_g": "0.19",
        "trans_fat_g": null,
        "fiber_g": "1.20",
        "sugar_g": "1.45",
        "cholesterol_mg": "1.00",
        "sodium_mg": "2588.00",
        "potassium_mg": "233.00",
        "vitamin_a_daily_pct": "12.00",
        "vitamin_c_daily_pct": "23.00",
        "calcium_daily_pct": "2.00",
        "iron_daily_pct": "8.00"
    },
    {
        "id": 3,
        "fatsecret_food_id": 6847,
        "serving_id": 23251,
        "retrieved_at": "2019-12-15T06:14:38.022Z",
        "food_name": "Vegetable Soup (Dry Mix)",
        "serving_url": "https://www.fatsecret.com/calories-nutrition/generic/vegetable-soup-dry-mix-not-reconstituted?portionid=23251&portionamount=1.000",
        "serving_desc": "1 Amount dry mix to make 8 fl oz prepared",
        "metric_serving_amt": "18.50",
        "metric_serving_unit": "g",
        "calories_kcal": "60.00",
        "fat_g": "0.94",
        "carbs_g": "11.08",
        "protein_g": "2.17",
        "saturated_fat_g": "0.43",
        "monounsaturated_fat_g": "0.35",
        "polyunsaturated_fat_g": "0.09",
        "trans_fat_g": null,
        "fiber_g": "0.60",
        "sugar_g": "0.70",
        "cholesterol_mg": "1.00",
        "sodium_mg": "1244.00",
        "potassium_mg": "112.00",
        "vitamin_a_daily_pct": "6.00",
        "vitamin_c_daily_pct": "11.00",
        "calcium_daily_pct": "1.00",
        "iron_daily_pct": "4.00"
    },
    {
        "id": 4,
        "fatsecret_food_id": 6847,
        "serving_id": 55527,
        "retrieved_at": "2019-12-15T06:14:38.022Z",
        "food_name": "Vegetable Soup (Dry Mix)",
        "serving_url": "https://www.fatsecret.com/calories-nutrition/generic/vegetable-soup-dry-mix-not-reconstituted?portionid=55527&portionamount=100.000",
        "serving_desc": "100 g",
        "metric_serving_amt": "100.00",
        "metric_serving_unit": "g",
        "calories_kcal": "325.00",
        "fat_g": "5.09",
        "carbs_g": "59.90",
        "protein_g": "11.73",
        "saturated_fat_g": "2.32",
        "monounsaturated_fat_g": "1.91",
        "polyunsaturated_fat_g": "0.50",
        "trans_fat_g": null,
        "fiber_g": "3.00",
        "sugar_g": "3.77",
        "cholesterol_mg": "3.00",
        "sodium_mg": "6722.00",
        "potassium_mg": "605.00",
        "vitamin_a_daily_pct": "31.00",
        "vitamin_c_daily_pct": "59.00",
        "calcium_daily_pct": "5.00",
        "iron_daily_pct": "21.00"
    }
];

describe("<FoodItem />", () => {
  test("FoodItem component", () => {
    const mockUseToasts = sinon.spy(useToasts);

    // the component file might not be using searchTerm from state
    // tests for that were not created yet
    store = _bigMockStore_({
      foodItemsReducer: {
        items: mockItems,
        item: mockFoodItem
      },
      dailyLog: {
        consumed: {
          caloriesConsumed: 0,
          fatsConsumed: 0,
          carbsConsumed: 0,
          proteinConsumed: 0
        },
        currentDate: "2019-12-15",
        currentTimeZone: "America/New_York"
      },
      firebase: {
        auth: {
          uid: "1234"
        },
        profile: {
          isLoaded: true
        }
      }
    });

    wrapper = mount(
      <MemoryRouter initialEntries={["food_item/search"]}>
        <Provider store={store}>
          <FoodItem match={{
            path: "food_item"
          }}/>
        </Provider>
      </MemoryRouter>
    );

    wrapper = mount(
      <MemoryRouter initialEntries={["food_item/view/6847"]}>
        <Provider store={store}>
          <ToastProvider number="5000">
            <FoodItem match={{
              path: "food_item"
            }}/>
          </ToastProvider>
        </Provider>
      </MemoryRouter>
    );
    
    
    expect(wrapper.find("#food-name").first().exists()).toBe(true);

    


    
    console.log(mockUseToasts.called);
    console.log(wrapper.debug());

  });
});
