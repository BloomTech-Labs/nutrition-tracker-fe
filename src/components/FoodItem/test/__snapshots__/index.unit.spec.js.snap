// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchPage/> Renders 1`] = `ShallowWrapper {}`;

exports[`<SearchResults/> Renders 1`] = `ReactWrapper {}`;
